source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc

```diff
@@ -1,28 +1,76 @@
 #include "extensions/filters/common/local_ratelimit/local_ratelimit_impl.h"
 
+#include "common/protobuf/utility.h"
+
 namespace Envoy {
 namespace Extensions {
 namespace Filters {
 namespace Common {
 namespace LocalRateLimit {
 
-LocalRateLimiterImpl::LocalRateLimiterImpl(const std::chrono::milliseconds fill_interval,
-                                           const uint32_t max_tokens,
-                                           const uint32_t tokens_per_fill,
-                                           Event::Dispatcher& dispatcher)
+LocalRateLimiterImpl::LocalRateLimiterImpl(
+    const std::chrono::milliseconds fill_interval, const uint32_t max_tokens,
+    const uint32_t tokens_per_fill, Event::Dispatcher& dispatcher,
+    const Envoy::Protobuf::RepeatedPtrField<
+        envoy::extensions::common::ratelimit::v3::LocalRateLimitDescriptor>& descriptors,
+    bool per_route)
     : fill_interval_(fill_interval), max_tokens_(max_tokens), tokens_per_fill_(tokens_per_fill),
       fill_timer_(fill_interval_ > std::chrono::milliseconds(0)
                       ? dispatcher.createTimer([this] { onFillTimer(); })
-                      : nullptr) {
+                      : nullptr),
+      time_source_(dispatcher.timeSource()) {
   if (fill_timer_ && fill_interval_ < std::chrono::milliseconds(50)) {
     throw EnvoyException("local rate limit token bucket fill timer must be >= 50ms");
   }
 
-  tokens_ = max_tokens;
+  tokens_.tokens = max_tokens;
 
   if (fill_timer_) {
     fill_timer_->enableTimer(fill_interval_);
   }
+
+  if (per_route) {
+    for (const auto& descriptor : descriptors) {
+      // If there is no limit or descriptors configured, then there is no config
+      // to rate limit on.
+      if (descriptor.entries().empty() || !descriptor.has_token_bucket()) {
+        throw EnvoyException(
+            "wrong config supplied for local rate limit. Either descriptor entries "
+            "or token bucket are missing.");
+        continue;
+      }
+
+      Envoy::RateLimit::LocalDescriptor new_descriptor;
+      for (const auto& entry : descriptor.entries()) {
+        new_descriptor.entries_.push_back({entry.key(), entry.value()});
+      }
+      Envoy::RateLimit::TokenBucket token_bucket;
+      token_bucket.fill_interval_ = absl::Milliseconds(
+          PROTOBUF_GET_MS_OR_DEFAULT(descriptor.token_bucket(), fill_interval, 0));
+      if (absl::ToChronoMilliseconds(token_bucket.fill_interval_).count() %
+              fill_interval_.count() !=
+          0) {
+        throw EnvoyException(
+            "local rate descriptor limit is not a multiple of token bucket fill timer");
+      }
```

Comment:
Is this logic coped from somewhere else? Share with the parent object logic for filling out token buckets, etc.

Reply:
We are filling token buckets here only now.. not shared with anything else..

Reply:
This doesn't make sense. This is inside the descriptor loop. The same error checking must happen for the global token bucket.

Reply:
we are making sure that descriptor token bucket fill interval is a multiple of global token bucket as the timer is on global token bucket

Reply:
Ah OK I see. Please document this somewhere on the actual proto fields. This is not clear from the docs.

Reply:
sure.. will fix

Reply:
Fixed